### ISSUE TITLE
Standardizing Hydra Schema

### ISSUE THREAD

**Issue by mostaphaRoudsari** [HydraShare/hydra#12]
@mdahlhausen, @chriswmackey and @alfarok and anybody else who is reading this. Here is a proposal to standardize Hydra schema before we get more users. The current schema is software specific and is not generic and extendable. To avoid ending up with slightly different schema for different platforms I'm proposing the structure below.

The main changed are:
- I merged all the image fields into a single images field
- I removed date and version fields as we should take them from the github repository. Having two versions and two time stamps can be pretty confusing.

``` javascript
{
"file" : /*full name to the file that will be downloaded on click on download button.
          the file should be located in root.*/
           "myporject.zip",

"thumbnail" : /* full name to an image which is located in root */
          "thumbnail.jpg",

"images" : [    /* 
                        List of images as objects: {image full name : optional description}
                        First image will be loaded on page load by default. All the images should be
                        located in the root folder
                        */
                        {"image1.jpg" : "Grasshopper definition"},
                        {"image2.gif": "Animated Rhino Scene"},
                        {"image3.png": "Dynamo definition"},
                         {"Image4.tiff": "", ...}
                   ],
"videos": {  /*
                  links to videos on youtube, Vimeo, etc.: optional description
                  */
                {"https://www.youtube.com/watch?v=9_u0dnyq2QI", "sunpath tutorial 1"}, ...
                },

"tags": [  /* list of tags as string */
              "EnergyPlus", "Honeybee", "Energy", ...
           ],

"components": { /* component/node name : number of occurrence in the file */
                          "Honeybee_Set EnergyPlus Zone Loads": 1,
                          "Honeybee_Construct Energy Balance": 1, ... 
                        },

"dependencies": [ /* list of plugins used in the definition */
                           "Human", "Ladybug", "Elk", "Mantis Shrimp", ...
                          ]
}
```
## Folder Structure

A Hydra folder includes:
- **README.md**: A markdown file that includes descriptions about the Hydra. The file name should always stay as README.md
-  **input.json**: A json file that follows the above structure.  The file name should always stay as input.json (all lower case)
- **thumbnail file**: An image file that is addressed in input.json as thumbnail
- **image files**: Any image file that is addressed in input.json
- **to be downloaded file**: The file that is addressed as "file" in input.json. This file is the only file that will be downloaded

**Comment by chriswmackey** [comment]
@mostaphaRoudsari ,

The structure generally looks good.  I will admit that I still do not fully understand how the images work but I trust that you have a solid understanding and will be able to implement it.  Let me know if you need any help navigating around the grasshopper component.  I tried to make the code as clean as possible.

-Chris

**Comment by mostaphaRoudsari** [comment]
@chriswmackey, I modified the component and web pages to work with the new schema. They still support the old schema from Grasshopper but you will see a warning that the example is generated by an older version of Hydra.

Back to images [this example](http://hydrashare.github.io/hydra/viewer?owner=mostaphaRoudsari&fork=hydra_1&id=shadow_range_study) might help to make it clear. As you can see now you can document your example with more than Rhino and Grasshopper images. Think as a Hydra page like a blog post. You may have some images prepared for a specific purpose. I think you should be able to have them included in Hydra. Also not all the example files will be generated from Rhino and Grasshopper. That's why I changed the field to images to keep it generic and compatible with other platforms.

Let me know if it looks good to you and we can close this issue.

**Comment by alfarok** [comment]
This looks good to me!  I think standardizing the platform is a good idea especially when the Dynamo Component is finished to help keep everything uniform.

**Comment by chriswmackey** [comment]
@mostaphaRoudsari ,

Excellent!  Your example looks really exciting.  I understand the general idea of allowing people to put up multiple images and trying to make the infrastructure not platform-specific.  However, I was just confused about what is happening on the back end as I see the images in your example are still named 'Rhino' and 'gh' on your fork:
https://github.com/mostaphaRoudsari/hydra_1/tree/master/shadow_range_study
To be more specific, is there some sort of naming scheme that we should follow if we want a certain image to show up first, second, third, etc. in the slide show?

-Chris

**Comment by mostaphaRoudsari** [comment]
@chriswmackey. Thanks. It's not about the file names but the file structure. I left the naming as what it used to be, but what is different between the previous structure and the current one is that all the images are under input.images - They can be named whatever that you want as far as they are available in the folder. In the other word there is no Rhino vs GH images. There are just couple of images.

Good point! There is no naming schema and first image will show up first. I think that's a fair point and we can change the schema to follow an order for images. Something like:

``` javascript
"images" : {     /* 
                        image full name : optional description
                        First image will be loaded by default. All the images should be
                        located in the root folder
                        */
                       "0" : { "image1.jpg" : "Grasshopper definition"},
                        "1": {"image2.gif": "Animated Rhino Scene"},
                        "2": {"image3.png": "Dynamo definition"},
                         "3": {"Image4.tiff": ""}, ...
                   },
```

or it can be simply a list of objects which will be in order. I can implement this change later tonight. It needs changes also in the front end.

**Comment by mdahlhausen** [comment]
@mostaphaRoudsari this looks good - makes it more extendable.  You may want to update the parser script on the viewer code to match the naming convention.  
Question on the image files - are we agnostic to image file types? do the order matter?

**Comment by chriswmackey** [comment]
@mostaphaRoudsari ,

Thanks for the explanation and it is much clearer now. To drive home my only major suggestions if we want people to use this feature: we should make it clear to them that the images should be in the folder BEFORE running the Hydra component and that the order of images that they want should follow a naming convention.

Also, how do you feel about having the GH script as the first image by default (unless people edit the json file)?  This sounds like a good means of shaping a general structure to the Hydra database while still allowing people to freedom to over-ride it (if they know what they are doing and edit the json).

-Chris

**Comment by mostaphaRoudsari** [comment]
@mdahlhausen yes we are agnostic to image types. I changed the images to a list of objects so now order does matter. First image on the list will show up first. I modified it in the first comment on this page.

@chriswmackey users don't need to copy the files manually (however they can if they want). I added a new input to the component to add additional images and component takes care of copying the files.

![image](https://cloud.githubusercontent.com/assets/2915573/10270019/b079b750-6ab3-11e5-9058-7b67de50ee20.png)

As I mentioned above, the order is now fixed so Grasshopper image will always show up first unless user modifies the json file manually. Here is the [same example](http://hydrashare.github.io/hydra/viewer?owner=mostaphaRoudsari&fork=hydra_1&id=shadow_range_study) file exported with the new component, and Grasshopper definition will be always loaded first.
